README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.org/logstash-plugins/logstash-output-elasticsearch_java.svg)](https://travis-ci.org/logstash-plugins/logstash-output-elasticsearch_java)

[![Travis Build Status](https://travis-ci.org/logstash-plugins/logstash-output-elasticsearch_java.svg)](https://travis-ci.org/logstash-plugins/logstash-output-elasticsearch_java)

[![Build Status](https://travis-ci.org/logstash-plugins/logstash-output-elasticsearch_java.svg)](https://travis-ci.org/logstash-plugins/logstash-output-elasticsearch_java)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run unit tests

```sh
bundle exec rspec
```

- Run integration tests

Dependencies: [Docker](http://docker.com)

Before the test suite is run, we will load and run an
Elasticsearch instance within a docker container. This container 
will be cleaned up when suite has finished.

```sh
bundle exec rspec --tag integration
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstash-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.


CHANGELOG.md
## 2.1.6
  - Docs: Set the default_codec doc attribute.

## 2.1.5
  - Fix some documentation issues

# 2.1.3
  - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash
# 2.1.2
  - New dependency requirements for logstash-core for the 5.0 release
## 2.1.1
 - Support Elasticsearch 2.1.0
 - Fix issue where transport client w/o port specified breaks
## 2.1.0
 - Port option no longer allowed
 - Majority of code now shared with logstash-output-elasticsearch
## 2.0.2
 - Update loading of addons to conform with new naming scheme (replaced - with _)
## 2.0.1
 - Make the default template doc_value aware
## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully, 
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0

## 2.0.0
  - Initial Release. Only supports Node/Transport in ES2.x

spec/es_spec_helper.rb
require "logstash/devutils/rspec/spec_helper"
require "ftw"
require "logstash/plugin"
require "logstash/json"
require "stud/try"
require "longshoreman"
require "logstash/outputs/elasticsearch_java"
require "logstash/outputs/elasticsearch_java/protocol"

CONTAINER_NAME = "logstash-output-elasticsearch-#{rand(999).to_s}"
CONTAINER_IMAGE = "elasticsearch"
CONTAINER_TAG = "1.6"

DOCKER_INTEGRATION = ENV["DOCKER_INTEGRATION"]

module ESHelper
  def get_local_host
    "127.0.0.1"
  end

  def get_host
    DOCKER_INTEGRATION ? Longshoreman.new.get_host_ip : "127.0.0.1"
  end

  def get_port(protocol)
    unless DOCKER_INTEGRATION
      return protocol.to_sym == :http ? 9200 : 9300
    end

    container = Longshoreman::Container.new
    container.get(CONTAINER_NAME)
    container.rport(9300)
  end

  def get_client
    Elasticsearch::Client.new(:hosts => "#{get_host}:#{get_port('http')}")
  end
end


RSpec.configure do |config|
  config.include ESHelper


  if DOCKER_INTEGRATION
    # this :all hook gets run before every describe block that is tagged with :integration => true.
    config.before(:all, :integration => true) do


      # check if container exists already before creating new one.
      begin
        ls = Longshoreman::new
        ls.container.get(CONTAINER_NAME)
      rescue Docker::Error::NotFoundError
        Longshoreman.new("#{CONTAINER_IMAGE}:#{CONTAINER_TAG}", CONTAINER_NAME)
        # TODO(talevy): verify ES is running instead of static timeout
        sleep 10
      end
    end

    # we want to do a final cleanup after all :integration runs,
    # but we don't want to clean up before the last block.
    # This is a final blind check to see if the ES docker container is running and
    # needs to be cleaned up. If no container can be found and/or docker is not
    # running on the system, we do nothing.
    config.after(:suite) do
      # only cleanup docker container if system has docker and the container is running
      begin
        ls = Longshoreman::new
        ls.container.get(CONTAINER_NAME)
        ls.cleanup
      rescue Docker::Error::NotFoundError, Excon::Errors::SocketError
        # do nothing
      end
    end
  end

  config.after(:each) do
    LogStash::Outputs::ElasticSearchJavaPlugins::Protocols::NodeClient.clear_client()
  end
end


spec/integration/outputs/index_spec.rb
require_relative "../../../spec/es_spec_helper"

shared_examples "an indexer" do
    let(:index) { 10.times.collect { rand(10).to_s }.join("") }
    let(:type) { 10.times.collect { rand(10).to_s }.join("") }
    let(:event_count) { 10000 + rand(500) }
    let(:flush_size) { rand(200) + 1 }
    let(:config) { "not implemented" }

    it "ships events" do
      insist { config } != "not implemented"

      pipeline = LogStash::Pipeline.new(config)
      pipeline.run

      index_url = "http://#{get_host}:#{get_port('http')}/#{index}"

      ftw = FTW::Agent.new
      ftw.post!("#{index_url}/_refresh")

      # Wait until all events are available.
      Stud::try(10.times) do
        data = ""
        response = ftw.get!("#{index_url}/_count?q=*")
        response.read_body { |chunk| data << chunk }
        result = LogStash::Json.load(data)
        cur_count = result["count"]
        insist { cur_count } == event_count
      end

      response = ftw.get!("#{index_url}/_search?q=*&size=1000")
      data = ""
      response.read_body { |chunk| data << chunk }
      result = LogStash::Json.load(data)
      result["hits"]["hits"].each do |doc|
        insist { doc["_type"] } == type
        insist { doc["_index"] } == index
      end
    end
end

describe "an indexer with custom index_type", :integration => true do
  it_behaves_like "an indexer" do
    let(:config) {
      <<-CONFIG
      input {
        generator {
          message => "hello world"
          count => #{event_count}
          type => "#{type}"
        }
      }
      output {
        elasticsearch_java {
          hosts => "#{get_host()}:#{get_port('transport')}"
          network_host => "#{get_local_host}"
          protocol => "transport"
          index => "#{index}"
          flush_size => #{flush_size}
        }
      }
      CONFIG
    }
  end
end

describe "an indexer with no type value set (default to logs)", :integration => true do
  it_behaves_like "an indexer" do
    let(:type) { "logs" }
    let(:config) {
      <<-CONFIG
      input {
        generator {
          message => "hello world"
          count => #{event_count}
        }
      }
      output {
        elasticsearch_java {
          hosts => "#{get_host()}:#{get_port('transport')}"
          network_host => "#{get_local_host}"
          protocol => "transport"
          index => "#{index}"
          flush_size => #{flush_size}
        }
      }
      CONFIG
    }
  end
end


spec/integration/outputs/retry_spec.rb
require "elasticsearch"
require "logstash/outputs/elasticsearch_java"
require_relative "../../../spec/es_spec_helper"

describe "failures in bulk class expected behavior", :integration => true do
  let(:template) { '{"template" : "not important, will be updated by :index"}' }
  let(:event1) { LogStash::Event.new("somevalue" => 100, "@timestamp" => "2014-11-17T20:37:17.223Z", "@metadata" => {"retry_count" => 0}) }
  let(:action1) { ["index", {:_id=>nil, :_routing=>nil, :_index=>"logstash-2014.11.17", :_type=>"logs"}, event1] }
  let(:event2) { LogStash::Event.new("geoip" => { "location" => [ 0.0, 0.0] }, "@timestamp" => "2014-11-17T20:37:17.223Z", "@metadata" => {"retry_count" => 0}) }
  let(:action2) { ["index", {:_id=>nil, :_routing=>nil, :_index=>"logstash-2014.11.17", :_type=>"logs"}, event2] }
  let(:invalid_event) { LogStash::Event.new("geoip" => { "location" => "notlatlon" }, "@timestamp" => "2014-11-17T20:37:17.223Z") }
  let(:max_retries) { 3 }

  def mock_actions_with_response(*resp)
    expanded_responses = resp.map do |resp|
      items = resp["statuses"] && resp["statuses"].map do |status|
        {"create" => {"status" => status, "error" => "Error for #{status}"}}
      end

      {
        "errors" => resp["errors"],
        "items" => items
      }
    end

    allow_any_instance_of(LogStash::Outputs::ElasticSearchJavaPlugins::Protocols::NodeClient).to receive(:bulk).and_return(*expanded_responses)
  end

  subject! do
    settings = {
      "manage_template" => true,
      "index" => "logstash-2014.11.17",
      "template_overwrite" => true,
      "network_host" => get_local_host,
      "protocol" => 'transport',
      "hosts" => "#{get_host()}:#{get_port('transport')}",
      "retry_max_items" => 10,
      "retry_max_interval" => 1,
      "max_retries" => max_retries
    }
    next LogStash::Outputs::ElasticSearchJava.new(settings)
  end

  before :each do
    # Clean ES of data before we start.
    @es = get_client
    @es.indices.delete_template(:name => "*")
    @es.indices.delete(:index => "*")
    @es.indices.refresh
  end

  it "should return no errors if all bulk actions are successful" do
    mock_actions_with_response({"errors" => false})
    expect(subject).to receive(:submit).with([action1, action2]).once.and_call_original
    subject.register
    subject.receive(event1)
    subject.receive(event2)
    subject.flush
    sleep(2)
  end

  it "retry exceptions within the submit body" do
    call_count = 0
    subject.register

    expect(subject.client).to receive(:bulk).with(anything).exactly(3).times do
      if (call_count += 1) <= 2
        raise "error first two times"
      else
        {"errors" => false}
      end
    end

    subject.receive(event1)
    subject.flush
  end

  it "should retry actions with response status of 503" do
    mock_actions_with_response({"errors" => true, "statuses" => [200, 200, 503, 503]},
                               {"errors" => true, "statuses" => [200, 503]},
                               {"errors" => false})
    expect(subject).to receive(:submit).with([action1, action1, action1, action2]).ordered.once.and_call_original
    expect(subject).to receive(:submit).with([action1, action2]).ordered.once.and_call_original
    expect(subject).to receive(:submit).with([action2]).ordered.once.and_call_original

    subject.register
    subject.receive(event1)
    subject.receive(event1)
    subject.receive(event1)
    subject.receive(event2)
    subject.flush
    sleep(3)
  end

  it "should retry actions with response status of 429" do
    mock_actions_with_response({"errors" => true, "statuses" => [429]},
                               {"errors" => false})
    expect(subject).to receive(:submit).with([action1]).twice.and_call_original
    subject.register
    subject.receive(event1)
    subject.flush
    sleep(3)
  end

  it "should retry an event until max_retries reached" do
    mock_actions_with_response({"errors" => true, "statuses" => [429]},
                               {"errors" => true, "statuses" => [429]},
                               {"errors" => true, "statuses" => [429]},
                               {"errors" => true, "statuses" => [429]},
                               {"errors" => true, "statuses" => [429]},
                               {"errors" => true, "statuses" => [429]})
    expect(subject).to receive(:submit).with([action1]).exactly(max_retries+1).times.and_call_original
    subject.register
    subject.receive(event1)
    subject.flush
    sleep(5)
  end

  it "non-retryable errors like mapping errors (400) should be dropped and not be retried (unfortunetly)" do
    subject.register
    subject.receive(invalid_event)
    expect(subject).not_to receive(:retry_push)
    subject.close

    @es.indices.refresh
    sleep(5)
    Stud::try(10.times) do
      r = @es.search
      insist { r["hits"]["total"] } == 0
    end
  end

  it "successful requests should not be appended to retry queue" do
    subject.register
    subject.receive(event1)
    expect(subject).to receive(:submit).once.and_call_original
    subject.close
    @es.indices.refresh
    sleep(5)
    Stud::try(10.times) do
      r = @es.search
      insist { r["hits"]["total"] } == 1
    end
  end

  it "should only index proper events" do
    subject.register
    subject.receive(invalid_event)
    subject.receive(event1)
    subject.close
    @es.indices.refresh
    sleep(5)

    Stud::try(10.times) do
      r = @es.search
      insist { r["hits"]["total"] } == 1
    end
  end
end


lib/logstash/outputs/elasticsearch_java/protocol.rb
require "cabin"
require "base64"
require 'logstash-output-elasticsearch_java_jars.rb'
require 'logstash/outputs/elasticsearch_java'

module LogStash module Outputs module ElasticSearchJavaPlugins module Protocols
  DEFAULT_OPTIONS = {
    :port => 9300,
  }

  class NodeClient
    attr_reader :settings, :client_options

    CLIENT_MUTEX = Mutex.new

    def initialize(options={})
      @logger = Cabin::Channel.get
      @client_options = DEFAULT_OPTIONS.merge(options)
      create_settings
    end

    def client_mutex_synchronize
      CLIENT_MUTEX.synchronize { yield }
    end

    def client
      client_mutex_synchronize { @@client ||= make_client }
    end

    # For use in test helpers
    def self.clear_node_client
      client_mutex_synchronize { @@client = nil }
    end

    def create_settings
      @settings = org.elasticsearch.common.settings.Settings.settingsBuilder()
      if @client_options[:hosts]
        @settings.put("discovery.zen.ping.multicast.enabled", false)
        @settings.put("discovery.zen.ping.unicast.hosts", hosts(@client_options))
      end

      @settings.put("node.client", true)
      @settings.put("http.enabled", false)
      @settings.put("path.home", Dir.pwd)

      if @client_options[:client_settings]
        @client_options[:client_settings].each do |key, value|
          @settings.put(key, value)
        end
      end

      @settings
    end

    def hosts(options)
      # http://www.elasticsearch.org/guide/reference/modules/discovery/zen/
      result = Array.new
      if options[:hosts].class == Array
        options[:hosts].each do |host|
          if host.to_s =~ /^.+:.+$/
            # For host in format: host:port, ignore options[:port]
            result << host
          else
            if options[:port].to_s =~ /^\d+-\d+$/
              # port ranges are 'host[port1-port2]'b
              result << Range.new(*options[:port].split("-")).collect { |p| "#{host}:#{p}" }
            else
              result << "#{host}:#{options[:port]}"
            end
          end
        end
      else
        if options[:hosts].to_s =~ /^.+:.+$/
          # For host in format: host:port, ignore options[:port]
          result << options[:hosts]
        else
          if options[:port].to_s =~ /^\d+-\d+$/
            # port ranges are 'host[port1-port2]' according to
            # http://www.elasticsearch.org/guide/reference/modules/discovery/zen/
            # However, it seems to only query the first port.
            # So generate our own list of unicast hosts to scan.
            range = Range.new(*options[:port].split("-"))
            result << range.collect { |p| "#{options[:hosts]}:#{p}" }
          else
            result << "#{options[:hosts]}:#{options[:port]}"
          end
        end
      end
      result.flatten.join(",")
    end

    # Normalizes the Java response to a reasonable approximation of the HTTP datastructure for interop
    # with the HTTP code
    def normalize_bulk_response(bulk_response)
      # TODO(talevy): parse item response objects to retrieve correct 200 (OK) or 201(created) status codes		+            items = bulk_response.map {|i|
      items = bulk_response.map { |i|
        if i.is_failed
          [[i.get_op_type, {"status" => i.get_failure.get_status.get_status, "message" => i.failureMessage}]]
        else
          [[i.get_op_type, {"status" => 200, "message" => "OK"}]]
        end
      }
      if bulk_response.has_failures()
        {"errors" => true, "items" => items}
      else
        {"errors" => false}
      end
    end

    def make_client
      nodebuilder = org.elasticsearch.node.NodeBuilder.nodeBuilder
      nodebuilder.settings(settings.build).node().client()
    end

    def bulk(actions)
      # Actions an array of [ action, action_metadata, source ]
      prep = client.prepareBulk
      actions.each do |action, args, source|
        prep.add(build_request(action, args, source))
      end
      response = prep.execute.actionGet()

      self.normalize_bulk_response(response)
    end

    # def bulk

    def build_request(action, args, source)
      case action
        when "index"
          request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
          request.id(args[:_id]) if args[:_id]
          request.routing(args[:_routing]) if args[:_routing]
          request.source(source)
        when "delete"
          request = org.elasticsearch.action.delete.DeleteRequest.new(args[:_index])
          request.id(args[:_id])
          request.routing(args[:_routing]) if args[:_routing]
        when "create"
          request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
          request.id(args[:_id]) if args[:_id]
          request.routing(args[:_routing]) if args[:_routing]
          request.source(source)
          request.opType("create")
        when "create_unless_exists"
          unless args[:_id].nil?
            request = org.elasticsearch.action.index.IndexRequest.new(args[:_index])
            request.id(args[:_id])
            request.routing(args[:_routing]) if args[:_routing]
            request.source(source)
            request.opType("create")
          else
            raise(LogStash::ConfigurationError, "Specifying action => 'create_unless_exists' without a document '_id' is not supported.")
          end
        when "update"
          unless args[:_id].nil?
            request = org.elasticsearch.action.update.UpdateRequest.new(args[:_index], args[:_type], args[:_id])
            request.routing(args[:_routing]) if args[:_routing]
            request.doc(source)
            if @client_options[:doc_as_upsert]
              request.docAsUpsert(true)
            else
              request.upsert(args[:_upsert]) if args[:_upsert]
            end
          else
            raise(LogStash::ConfigurationError, "Specifying action => 'update' without a document '_id' is not supported.")
          end
        else
          raise(LogStash::ConfigurationError, "action => '#{action_name}' is not currently supported.")
      end # case action

      request.type(args[:_type]) if args[:_type]
      return request
    end

    # def build_request

    def template_exists?(name)
      return !client.admin.indices.
        prepareGetTemplates(name).
        execute().
        actionGet().
        getIndexTemplates().
        isEmpty
    end

    def template_install(name, template, force=false)
      if template_exists?(name) && !force
        @logger.debug("Found existing Elasticsearch template. Skipping template management", :name => name)
        return
      end
      template_put(name, template)
    end

    def template_put(name, template)
      response = client.admin.indices.
        preparePutTemplate(name).
        setSource(LogStash::Json.dump(template)).
        execute().
        actionGet()

      raise "Could not index template!" unless response.isAcknowledged
    end
  end # class NodeClient

  class TransportClient < NodeClient
    private
    def make_client
      builder = org.elasticsearch.client.transport.TransportClient.builder()
      client = client_options[:elasticsearch_plugins].reduce(builder) do |b,plugin_class|
        b.add_plugin(plugin_class)
      end.settings((settings.build)).build()

      client_options[:hosts].each do |host|
        matches = host.match /([^:+]+)(:(\d+))?/

        inet_addr = java.net.InetAddress.getByName(matches[1])
        port = (matches[3] || 9300).to_i
        client.addTransportAddress(
          org.elasticsearch.common.transport.InetSocketTransportAddress.new(
            inet_addr, port
          )
        )
      end

      return client
    end

    # We want a separate client per instance for transport
    def client
      client_mutex_synchronize { @client ||= make_client }
    end

    def clear_client()
      client_mutex_synchronize { @client = nil }
    end
  end
end end end end

lib/logstash/outputs/elasticsearch_java.rb
# encoding: utf-8
require "logstash/namespace"
require "logstash/environment"
require "logstash/outputs/base"
require "logstash/json"
require "concurrent"
require "socket" # for Socket.gethostname
require "thread" # for safe queueing
require "uri" # for escaping user input
require "logstash/outputs/elasticsearch_java/protocol"
require "logstash/outputs/elasticsearch"

# IMPORTANT: This plugin is no longer supported. Please use the recommended
# <<plugins-outputs-elasticsearch,Elasticsearch output plugin>> instead.
#
#This output lets you store logs in Elasticsearch using the native 'node' and 'transport'
# protocols. It is highly recommended to use the regular 'logstash-output-elasticsearch' output
# which uses HTTP instead. This output is, in-fact, sometimes slower, and never faster than that one.
# Additionally, upgrading your Elasticsearch cluster may require you to simultaneously update this
# plugin for any protocol level changes. The HTTP client may be easier to work with due to wider
# familiarity with HTTP.
#
# *VERSION NOTE*: Your Elasticsearch cluster must be running Elasticsearch 1.0.0 or later.
#
# If you want to set other Elasticsearch options that are not exposed directly
# as configuration options, there are two methods:
#
# * Create an `elasticsearch.yml` file in the $PWD of the Logstash process
# * Pass in es.* java properties (`java -Des.node.foo=` or `ruby -J-Des.node.foo=`)
#
# With the default `protocol` setting ("node"), this plugin will join your
# Elasticsearch cluster as a client node, so it will show up in Elasticsearch's
# cluster status.
#
# You can learn more about Elasticsearch at <https://www.elastic.co/products/elasticsearch>
#
# ==== Operational Notes
#
# If using the default `protocol` setting ("node"), your firewalls might need
# to permit port 9300 in *both* directions (from Logstash to Elasticsearch, and
# Elasticsearch to Logstash)
#
# ==== Retry Policy
#
# By default all bulk requests to ES are synchronous. Not all events in the bulk requests
# always make it successfully. For example, there could be events which are not formatted
# correctly for the index they are targeting (type mismatch in mapping). So that we minimize loss of 
# events, we have a specific retry policy in place. We retry all events which fail to be reached by 
# Elasticsearch for network related issues. We retry specific events which exhibit errors under a separate 
# policy described below. Events of this nature are ones which experience ES error codes described as 
# retryable errors.
#
# *Retryable Errors:*
#
# - 429, Too Many Requests (RFC6585)
# - 503, The server is currently unable to handle the request due to a temporary overloading or maintenance of the server.
# 
# Here are the rules of what is retried when:
#
# - Block and retry all events in bulk response that experiences transient network exceptions until
#   a successful submission is received by Elasticsearch.
# - Retry subset of sent events which resulted in ES errors of a retryable nature which can be found 
#   in RETRYABLE_CODES
# - For events which returned retryable error codes, they will be pushed onto a separate queue for 
#   retrying events. events in this queue will be retried a maximum of 5 times by default (configurable through :max_retries). The size of 
#   this queue is capped by the value set in :retry_max_items.
# - Events from the retry queue are submitted again either when the queue reaches its max size or when
#   the max interval time is reached, which is set in :retry_max_interval.
# - Events which are not retryable or have reached their max retry count are logged to stderr.
class LogStash::Outputs::ElasticSearchJava < LogStash::Outputs::Base
  attr_reader :client

  include LogStash::Outputs::ElasticSearch::CommonConfigs
  include LogStash::Outputs::ElasticSearch::Common

  RETRYABLE_CODES = [409, 429, 503]
  SUCCESS_CODES = [200, 201]

  config_name "elasticsearch_java"

  # The Elasticsearch action to perform. Valid actions are:
  #
  # - index: indexes a document (an event from Logstash).
  # - delete: deletes a document by id (An id is required for this action)
  # - create: indexes a document, fails if a document by that id already exists in the index.
  # - update: updates a document by id. Update has a special case where you can upsert -- update a
  #   document if not already present. See the `upsert` option
  # - create_unless_exists: create the document unless it already exists, in which case do nothing.
  #
  # For more details on actions, check out the http://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html[Elasticsearch bulk API documentation]
  config :action, :validate => %w(index delete create update create_unless_exists), :default => "index"

  # The name of your cluster if you set it on the Elasticsearch side. Useful
  # for discovery when using `node` or `transport` protocols.
  # By default, it looks for a cluster named 'elasticsearch'.
  # Equivalent to the Elasticsearch option 'cluster.name'
  config :cluster, :validate => :string

  # This sets the local port to bind to. Equivalent to the Elasticsrearch option 'transport.tcp.port'
  config :transport_tcp_port, :validate => :number

  # This setting no longer does anything. It exists to keep config validation
  # from failing. It will be removed in future versions.
  config :max_inflight_requests, :validate => :number, :default => 50, :deprecated => true

  # The node name Elasticsearch will use when joining a cluster.
  #
  # By default, this is generated internally by the ES client.
  config :node_name, :validate => :string

  # Choose the protocol used to talk to Elasticsearch.
  #
  # The 'node' protocol (default) will connect to the cluster as a normal Elasticsearch
  # node (but will not store data). If you use the `node` protocol, you must permit
  # bidirectional communication on the port 9300 (or whichever port you have
  # configured).
  #
  # If you do not specify the `host` parameter, it will use  multicast for http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-discovery-zen.html[Elasticsearch discovery].  While this may work in a test/dev environment where multicast is enabled in 
  # Elasticsearch, we strongly recommend http://www.elastic.co/guide/en/elasticsearch/guide/current/important-configuration-changes.html#unicast[using unicast]
  # in Elasticsearch.  To connect to an Elasticsearch cluster with unicast,
  # you must include the `host` parameter (see relevant section above).  
  #
  # The 'transport' protocol will connect to the host you specify and will
  # not show up as a 'node' in the Elasticsearch cluster. This is useful
  # in situations where you cannot permit connections outbound from the
  # Elasticsearch cluster to this Logstash server.
  #
  # All protocols will use bulk requests when talking to Elasticsearch.
  config :protocol, :validate => [ "node", "transport"], :default => "transport"

  # Enable cluster sniffing (transport only).
  # Asks host for the list of all cluster nodes and adds them to the hosts list
  # Equivalent to the Elasticsearch option 'client.transport.sniff'
  config :sniffing, :validate => :boolean, :default => false

  # The name/address of the host to bind to for Elasticsearch clustering. Equivalent to the Elasticsearch option 'network.host'
  # option.
  # This MUST be set for either protocol to work (node or transport)! The internal Elasticsearch node
  # will bind to this ip. This ip MUST be reachable by all nodes in the Elasticsearch cluster
  config :network_host, :validate => :string, :required => true

  def client_options
    client_settings = {}
    client_settings["cluster.name"] = @cluster if @cluster
    client_settings["network.host"] = @network_host if @network_host
    client_settings["transport.tcp.port"] = @transport_tcp_port if @transport_tcp_port
    client_settings["client.transport.sniff"] = @sniffing

    if @node_name
      client_settings["node.name"] = @node_name
    else
      client_settings["node.name"] = "logstash-#{Socket.gethostname}-#{$$}-#{object_id}"
    end

    options = {
      :protocol => @protocol,
      :client_settings => client_settings,
      :hosts => @hosts
    }

    # Update API setup
    update_options = {
      :upsert => @upsert,
      :doc_as_upsert => @doc_as_upsert
    }
    options.merge! update_options if @action == 'update'

    options
  end

  def build_client
    @client = client_class.new(client_options)
  end

  def close
    @stopping.make_true
    @buffer.stop
  end

  def get_plugin_options
    @@plugins.each do |plugin|
      name = plugin.name.split('-')[-1]
      client_settings.merge!(LogStash::Outputs::ElasticSearchJava.const_get(name.capitalize).create_client_config(self))
    end
  end

  def client_class
    case @protocol
      when "transport"
        LogStash::Outputs::ElasticSearchJavaPlugins::Protocols::TransportClient
      when "node"
        LogStash::Outputs::ElasticSearchJavaPlugins::Protocols::NodeClient
    end
  end

  @@plugins = Gem::Specification.find_all{|spec| spec.name =~ /logstash-output-elasticsearch_java_/ }

  @@plugins.each do |plugin|
    name = plugin.name.split('_')[-1]
    require "logstash/outputs/elasticsearch_java/#{name}"
  end

end # class LogStash::Outputs::ElasticSearchJava
